Remove or document Unsafe.AsPointer uses in core libraries

hamarb123 opened the pull request:
Contributes to https://github.com/dotnet/runtime/issues/99144

hamarb123 committed:
Remove or document Unsafe.AsPointer uses in System.Memory

- Also add internal helper method Unsafe.IsOpportunisticallyAligned
- And cast a pointer to IntPtr instead of int for a unit test

hamarb123 committed:
Fix typo and cast size to unsigned

huoyaoyuan commented on src/libraries/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.T.cs:
```diff
@@ -424,14 +424,7 @@ public void MultipleCallsToGetSpan()
-                    Assert.Equal((IntPtr)pSpan, (IntPtr)pNewSpan);
-                }
-
+                Assert.Equal(0, Unsafe.ByteOffset(ref MemoryMarshal.GetReference(span), ref MemoryMarshal.GetReference(newSpan)));
```
`Unsafe.AreSameRef`?

hamarb123 replied:
I can do `Unsafe.AreSame`, but I figured this might be better for debugging if it does fail. Will see what others' opinions are, but can change it to that if that's what we want.

hamarb123 replied:
@jkotas what's your opinion about this?

jkotas replied:
Either way is fine with me. I do not have a strong opinion.

> I figured this might be better for debugging if the test does fail.

I think it is very unlikely that this test is ever going to fail. We can always add more verbose diagnostic if we see the test failing intermittently.

jkotas commented on src/libraries/System.Memory/tests/MemoryMarshal/CreateFromPinnedArray.cs:
```diff
@@ -190,15 +190,16 @@ public static unsafe void CreateFromPinnedArrayVerifyPinning()
-            Assert.Equal((int)pinnedPtr, (int)Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedMemory.Span)));
-            Assert.Equal((int)memoryHandlePinnedPtr, (int)pinnedGCHandle.AddrOfPinnedObject().ToPointer());
+            Assert.Equal((IntPtr)pinnedPtr, (IntPtr)Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedMemory.Span)));
+            Assert.Equal((IntPtr)memoryHandlePinnedPtr, (IntPtr)pinnedGCHandle.AddrOfPinnedObject().ToPointer());
```
```suggestion
            Assert.Equal((IntPtr)memoryHandlePinnedPtr, pinnedGCHandle.AddrOfPinnedObject());
```

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:
```diff
@@ -907,5 +908,20 @@ public static void SkipInit<T>(out T value)
+
+        // Determines if the address is aligned at least to `alignment` bytes.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static bool IsOpportunisticallyAligned<T>(ref readonly T address, nuint alignment)
```
There are more places in CoreLib that should use this method.

hamarb123 replied:
I've only done System.Memory so far, as suggested by Tanner (to break it up by area). Did you want me to do the whole repo (this would be my preferred option), or a larger portion at once (let me know how much if this)?

jkotas replied:
I have seen ~5 places where this helper can be used. I think all of them can be fixed up in this PR.

Outside CoreLib, I have only seen one hit System.Text.Json where the shared helper cannot be used without introducing a lot of complexity.

hamarb123 replied:
So, update the other places to use alignment helper, but update the rest of the code in them separately? Or just do all the core library code?

tannergooding replied:
> as suggested by Tanner (to break it up by area, or at least that's how I understood it)

I mostly meant this as a way to help scope the changes based on relation to eachother. If the PR contains changes to S.T.Encoding and S.T.Json for example, it might be tougher to get in.

In this case, you're introducing a new API and updating places to use it en masse, so it's fine since that's all directly related to the new internal API, especially given there's only 4-5 places.

MichalPetryka commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:
```diff
@@ -907,5 +908,20 @@ public static void SkipInit<T>(out T value)
+            // GC will keep alignment when moving objects (up to sizeof(void*)),
+            // otherwise alignment should be considered a hint if not pinned.
+            Debug.Assert(nuint.IsPow2(alignment));
+            return ((nuint)AsPointer(ref AsRef(in address)) & (alignment - 1)) != 0;
```
```suggestion
            return ((nuint)AsPointer(ref AsRef(in address)) & (alignment - 1)) == 0;
```

hamarb123 replied:
that was clever of me lol ;)

> NOTE: jit can optimize `x % POT_CNS == 0` to this itself

I usually write it the current way (because I think it's more logical to think about which bits should be 0), but I can change it to this if we think it's more readable.

hamarb123 committed:
Address feedback

hamarb123 committed:
Changes for the rest of the core libraries

- Other than MethodHandle code for NativeAOT
- Add Unsafe.OpportunisticMisalignment

hamarb123 commented:
I've skipped the code in NativeAOT's `MethodHandle` files, and `QCallHandles`

hamarb123 committed:
Fix compile errors

hamarb123 committed:
Update CorInfoImpl.cs

MichalPetryka commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/NativeRuntimeEventSource.Threading.NativeSinks.cs:
```diff
@@ -355,12 +355,14 @@ public unsafe void ThreadPoolIOPack(NativeOverlapped* nativeOverlapped)
             WaitHandleWaitSourceMap waitSource = WaitHandleWaitSourceMap.Unknown,
             object? associatedObject = null) =>
-            WaitHandleWaitStart(waitSource, *(nint*)Unsafe.AsPointer(ref associatedObject));
+            WaitHandleWaitStart(waitSource, *(nint*)&associatedObject);
```
Would `Unsafe.As<object, nint>(ref associatedObject)` maybe look cleaner in cases like this?

jkotas replied:
EventSource uses these unpinned object pointers in number of places. 

It may be nice to create an internal helper for this on EventSource and switch all places to call it. `internal static long ObjectIDForEvents(object o)`

MichalPetryka replied:
> EventSource uses these unpinned object pointers in number of places.

This isn't an unpinned pointer though but effectively a `BitCast<object, nint>`.

jkotas replied:
NativeRuntimeEventSource.Threading.cs, Lock.cs and TplEventSource.cs are the other files that can use this helper.

For Lock.cs, I would delete the property and just call the internal helper where the event is emitted.

hamarb123 committed:
Apply feedback

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:
```diff
@@ -907,5 +908,37 @@ public static void SkipInit<T>(out T value)
+        // Returns the object as a IntPtr - safe when object is pinned, or when only used for logging;
+        // The second method is also defined so that safe usage is always obvious by name.
+        internal static nint GetPinnedObjectPointer(object o) =>  *(nint*)&o;
+        internal static nint ObjectIDForEvents(object o) => GetPinnedObjectPointer(o);
```
This method is specific to EventSource. I do not think it belongs here.

jkotas commented on src/coreclr/nativeaot/Common/src/Internal/Runtime/CompilerHelpers/StartupCodeHelpers.cs:
```diff
@@ -233,8 +234,7 @@ private static unsafe object[] InitializeStatics(IntPtr gcStaticRegionStart, int
                     // Update the base pointer to point to the pinned object
-                    *pBlock = *(IntPtr*)&obj;
-#pragma warning restore CS8500
+                    *pBlock = Unsafe.GetPinnedObjectPointer(obj);
```
I do not think that this is going to build.

This is very special low-level code. I think what you have here earlier was better.

hamarb123 committed:
Apply feedback

hamarb123 committed:
Apply feedback

hamarb123 committed:
Update Unsafe.cs

MichalPetryka commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs:
```diff
@@ -907,5 +908,38 @@ public static void SkipInit<T>(out T value)
+        }
+
+        // Returns the object as a IntPtr - safe when object is pinned, or when only used for logging;
+        internal static nint GetPinnedObjectPointer(object o)
```
Would `nuint` be better here?

hamarb123 replied:
Just deleted it anyway 😂. All uses used `nint`/`IntPtr` before though so I had it as that

jkotas replied:
All existing EventSource events log these IDs as signed int.

jkotas commented:
> Should I include changes for the misc tests in this PR too?

I would start a new PR for these.

hamarb123 committed:
Fix compile errors :)

sharwell commented on src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:
```diff
@@ -865,7 +865,9 @@ private static unsafe void ConfigCallback(void* configurationContext, void* name
             };
 
-            _EnumerateConfigurationValues(Unsafe.AsPointer(ref context), &ConfigCallback);
+#pragma warning disable CS8500
```
💡 As an aid to the reader, can these be updated to include the title for CS8500?

sharwell replied:
My request would only be for the ones in this PR. I'm not a maintainer on this repository, so I can't speak to the preferred comment if there is more than one form seen in other code.

hamarb123 committed:
Add full CS8500 comment & fix compile error

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs:
```diff
@@ -865,7 +865,9 @@ private static unsafe void ConfigCallback(void* configurationContext, void* name
-            _EnumerateConfigurationValues(Unsafe.AsPointer(ref context), &ConfigCallback);
+#pragma warning disable CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
+            _EnumerateConfigurationValues(&context, &ConfigCallback);
+#pragma warning restore CS8500 // This takes the address of, gets the size of, or declares a pointer to a managed type
```
The comment on pragma warning restore is not very useful.

hamarb123 replied:
That is the standard comment generated by VS

hamarb123 replied:
Do you want me to customise each one, or should we just use the standard comment?

jkotas replied:
I would delete the comment on `#pragma warning restore`. It is duplicate of the comment one  `#pragma warning disable` that is typically a few lines above, so it just a clutter with no real value. Vast majority of  `#pragma warning restore` do not have a comment like that.

jkotas replied:
> Do you want me to customise each one, or should we just use the standard comment?

I do not think it needs to be customized. 

My personal preference would be a short succinct comment. Long boiler plate comments are less useful. For example this looks good to me:

https://github.com/dotnet/runtime/blob/445a3e8a5e5c98a0e9fe1128ac742bf56f78277c/src/libraries/System.Private.Uri/src/System/UriHelper.cs#L15-L21

hamarb123 committed:
Address feedback

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TplEventSource.cs:
```diff
@@ -501,17 +501,21 @@ public void TraceSynchronousWorkEnd(CausalitySynchronousWork Work)
+#pragma warning disable CS8500 // takes address of managed type
         [NonEvent]
-        public unsafe void RunningContinuation(int TaskID, object Object) { RunningContinuation(TaskID, (long)*((void**)Unsafe.AsPointer(ref Object))); }
+        public unsafe void RunningContinuation(int TaskID, object Object) { RunningContinuation(TaskID, (long)*((void**)&Object)); }
```
This should use the `ObjectIDForEvents` method

hamarb123 committed:
Update TplEventSource.cs

jkotas reviewed (approved):
Thanks!
